### PR TITLE
Graylog index sets module for interacting with the Graylog API

### PR DESCRIPTION
**Pull request by shortstack**
Graylog index sets module for interacting with the Graylog API

##### SUMMARY
New module for interacting with the [Graylog](https://github.com/Graylog2/graylog2-server) API, specifically creating and manipulating index sets, which has been beneficial in automating Graylog functionality within CI pipelines/processes, since these have typically been manual processes.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
- graylog_index_sets

### CODE REVIEW

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
+#!/usr/bin/python
+# (c) 2019, Whitney Champion <whitney.ellis.champion@gmail.com>
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
# Copyright: (c) 2019, Whitney Champion <whitney.ellis.champion@gmail.com>
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+module: graylog_index_sets
+short_description: Communicate with the Graylog API to manage index sets
+description:
+    - The Graylog index sets module manages Graylog index sets
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
    - The Graylog index sets module manages Graylog index sets.
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+version_added: "2.9"
+author: "Whitney Champion (@shortstack)"
+options:
+  endpoint:
+    description:
+      - Graylog endoint. (i.e. graylog.mydomain.com).
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:23]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+  graylog_user:
+    description:
+      - Graylog privileged user username.
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:28]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+  shards:
+    description:
+      - Number of Elasticsearch shards used per index in this index set.
+    required: false
+    default: 4
+  replicas:
+    description:
+      - Number of Elasticsearch replicas used per index in this index set.
+    required: false
+    default: 1
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:90]:
```suggestion
    default: 1
    type: int
```

Comment (Akasurde) [reply lib/ansible/modules/monitoring/graylog_index_sets.py:90]: Match type for other parameters too.

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+  index_optimization_max_num_segments:
+    description:
+      - Maximum number of segments per Elasticsearch index after optimization (force merge).
+    required: false
+    default: 1
+  index_optimization_disabled:
+    description:
+      - Disable Elasticsearch index optimization (force merge) after rotation.
+    required: false
+    default: False
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:127]:
```suggestion
    default: False
    type: bool
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+
+# import module snippets
+import json
+import datetime
+import base64
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.urls import fetch_url, to_text
+
+
+def create(module, base_url, headers, title, description, index_prefix, index_analyzer, shards, replicas,
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
def create(module, base_url, headers):
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+           rotation_strategy_class, retention_strategy_class, rotation_strategy, retention_strategy,
+           index_optimization_max_num_segments, index_optimization_disabled, creation_date, writable, default):
+
+    url = base_url
+
+    payload = {}
+
+    if title is not None:
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]: How about - 

```suggestion
    for key in ['title', 'description', 'index_prefix', 'creation_date', 'writable', 'default', 'index_analyzer', 'shards', 'replicas', ]
        if module.params[key] is not None:
            payload[key] = module.params[key]
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+    if index_optimization_disabled is not None:
+        payload['index_optimization_disabled'] = index_optimization_disabled
+
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))
+
+    if info['status'] != 200:
+        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
+
+    try:
+        content = response.read()
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
        content = to_text(response.read(), errors='surrogate_or_strict')
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+    except AttributeError:
+        content = info.pop('body', '')
+
+    return info['status'], info['msg'], content, url
+
+
+def update(module, base_url, headers, index_set_id, title, description, index_prefix, index_analyzer, shards, replicas,
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]: Same here as `create` method

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+           rotation_strategy_class, retention_strategy_class, rotation_strategy, retention_strategy, index_optimization_max_num_segments,
+           index_optimization_disabled, writable, default):
+
+    url = base_url + "/%s" % (index_set_id)
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
    url = "/".join([base_url, index_set_id])
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+
+    payload = {}
+
+    if title is not None:
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]: Same here as `create` method

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 13]
```diff
@@ -0,0 +1,455 @@
...
+    if index_optimization_disabled is not None:
+        payload['index_optimization_disabled'] = index_optimization_disabled
+
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='PUT', data=module.jsonify(payload))
+
+    if info['status'] != 200:
+        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
+
+    try:
+        content = response.read()
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
        content = to_text(response.read(), errors='surrogate_or_strict')
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 14]
```diff
@@ -0,0 +1,455 @@
...
+
+    url = base_url + "/%s" % (index_set_id)
+
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='DELETE')
+
+    if info['status'] != 204:
+        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
+
+    try:
+        content = response.read()
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
        content = to_text(response.read(), errors='surrogate_or_strict')
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 15]
```diff
@@ -0,0 +1,455 @@
...
+    else:
+        url = base_url
+
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='GET')
+
+    if info['status'] != 200:
+        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
+
+    try:
+        content = response.read()
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
        content = to_text(response.read(), errors='surrogate_or_strict')
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 16]
```diff
@@ -0,0 +1,455 @@
...
+
+    url = base_url
+
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='GET')
+
+    if info['status'] != 200:
+        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
+
+    try:
+        content = response.read()
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
        content = to_text(response.read(), errors='surrogate_or_strict')
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 17]
```diff
@@ -0,0 +1,455 @@
...
+    return index_set_id
+
+
+def get_token(module, endpoint, username, password):
+
+    headers = '{ "Content-Type": "application/json", "X-Requested-By": "Graylog API", "Accept": "application/json" }'
+
+    url = "https://%s/api/system/sessions" % (endpoint)
+
+    payload = {}
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
    payload = {
      'username': username,
      'password': password,
      'host': endpoint
    }
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 18]
```diff
@@ -0,0 +1,455 @@
...
+    payload['password'] = password
+    payload['host'] = endpoint
+
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))
+
+    if info['status'] != 200:
+        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
+
+    try:
+        content = response.read()
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
        content = to_text(response.read(), errors='surrogate_or_strict')
```

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+    type: str
+  graylog_user:
+    description:
+      - Graylog privileged user username.
+    required: false
+    type: str
+  graylog_password:
+    description:
+      - Graylog privileged user password.
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:33]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+  action:
+    description:
+      - Action to take against index API.
+    required: false
+    default: list
+    choices: [ create, update, list, delete, query_index_sets ]
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:40]:
```suggestion
    choices: [ create, update, list, delete, query_index_sets ]
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+  index_set_id:
+    description:
+      - Index id.
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:45]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 22]
```diff
@@ -0,0 +1,405 @@
...
+  title:
+    description:
+      - Title.
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:50]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 23]
```diff
@@ -0,0 +1,405 @@
...
+  description:
+    description:
+      - Description.
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:55]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 24]
```diff
@@ -0,0 +1,405 @@
...
+  index_prefix:
+    description:
+      - A unique prefix used in Elasticsearch indices belonging to this index set. The prefix must start
+        with a letter or number, and can only contain letters, numbers, '_', '-' and '+'.
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:66]:
```suggestion
    required: false
```
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 25]
```diff
@@ -0,0 +1,405 @@
...
+  index_analyzer:
+    description:
+      - Elasticsearch analyzer for this index set.
+    required: false
+    default: "standard"
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:78]:
```suggestion
    default: "standard"
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 26]
```diff
@@ -0,0 +1,405 @@
...
+  shards:
+    description:
+      - Number of Elasticsearch shards used per index in this index set.
+    required: false
+    default: 4
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:84]:
```suggestion
    default: 4
    type: int
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 27]
```diff
@@ -0,0 +1,405 @@
...
+    description:
+      - Number of Elasticsearch replicas used per index in this index set.
+    required: false
+    default: 1
+    type: int
+  rotation_strategy_class:
+    description:
+      - Rotation strategy class, ex. org.graylog2.indexer.rotation.strategies.TimeBasedRotationStrategy
+    required: false
+    default: "org.graylog2.indexer.rotation.strategies.TimeBasedRotationStrategy"
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:96]:
```suggestion
    default: "org.graylog2.indexer.rotation.strategies.TimeBasedRotationStrategy"
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 28]
```diff
@@ -0,0 +1,405 @@
...
+  retention_strategy_class:
+    description:
+      - Retention strategy class, ex. org.graylog2.indexer.retention.strategies.DeletionRetentionStrategy
+    required: false
+    default: "org.graylog2.indexer.retention.strategies.DeletionRetentionStrategy"
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:102]:
```suggestion
    default: "org.graylog2.indexer.retention.strategies.DeletionRetentionStrategy"
    type: str
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 29]
```diff
@@ -0,0 +1,405 @@
...
+  rotation_strategy:
+    description:
+      - Graylog uses multiple indices to store documents in. You can configure the strategy it uses to determine
+         when to rotate the currently active write index.
+    required: false
+    default: {'type': 'org.graylog2.indexer.rotation.strategies.TimeBasedRotationStrategyConfig', 'rotation_period': 'P1D'}
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:109]:
```suggestion
    default: {'type': 'org.graylog2.indexer.rotation.strategies.TimeBasedRotationStrategyConfig', 'rotation_period': 'P1D'}
    type: dict
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 30]
```diff
@@ -0,0 +1,405 @@
...
+  retention_strategy:
+    description:
+      - Graylog uses a retention strategy to clean up old indices.
+    required: false
+    default: {'type': 'org.graylog2.indexer.retention.strategies.DeletionRetentionStrategyConfig', 'max_number_of_indices': 14}
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:115]:
```suggestion
    default: {'type': 'org.graylog2.indexer.retention.strategies.DeletionRetentionStrategyConfig', 'max_number_of_indices': 14}
    type: dict
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 31]
```diff
@@ -0,0 +1,405 @@
...
+  index_optimization_max_num_segments:
+    description:
+      - Maximum number of segments per Elasticsearch index after optimization (force merge).
+    required: false
+    default: 1
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:121]:
```suggestion
    default: 1
    type: int
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 32]
```diff
@@ -0,0 +1,405 @@
...
+  writable:
+    description:
+      - Writable, true or false.
+    required: false
+    default: True
+  default:
+    description:
+      - Default index set, true or false.
+    required: false
+    default: False
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:139]:
```suggestion
    default: False
    type: bool
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 33]
```diff
@@ -0,0 +1,405 @@
...
+    description:
+      - Disable Elasticsearch index optimization (force merge) after rotation.
+    required: false
+    default: False
+    type: bool
+  writable:
+    description:
+      - Writable, true or false.
+    required: false
+    default: True
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:133]:
```suggestion
    default: True
    type: bool
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 34]
```diff
@@ -0,0 +1,405 @@
...
+    payload = {}
+
+    for key in ['title', 'description', 'index_prefix', 'creation_date', 'writable', 'default',
+                'index_analyzer', 'shards', 'replicas', 'rotation_strategy_class', 'retention_strategy_class',
+                'rotation_strategy', 'retention_strategy', 'index_optimization_max_num_segments',
+                'index_optimization_disabled']:
+        if module.params[key] is not None:
+            payload[key] = module.params[key]
+
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:295]:
```suggestion
    response, info = fetch_url(module=module, url=base_url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 35]
```diff
@@ -0,0 +1,405 @@
...
+
+    if info['status'] != 200:
+        module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
+
+    try:
+        content = to_text(response.read(), errors='surrogate_or_strict')
+    except AttributeError:
+        content = info.pop('body', '')
+
+    return info['status'], info['msg'], content, url
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:305]:
```suggestion
    return info['status'], info['msg'], content, base_url
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 36]
```diff
@@ -0,0 +1,405 @@
...
+
+
+def query_index_sets(module, base_url, headers, title):
+
+    url = base_url
+
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='GET')
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:375]:
```suggestion
    response, info = fetch_url(module=module, url=base_url, headers=json.loads(headers), method='GET')
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 37]
```diff
@@ -0,0 +1,405 @@
...
+        )
+    )
+
+    endpoint = module.params['endpoint']
+    graylog_user = module.params['graylog_user']
+    graylog_password = module.params['graylog_password']
+    action = module.params['action']
+    title = module.params['title']
+    index_set_id = module.params['index_set_id']
+    description = module.params['description']
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]: Line 355 to 368 are unused.

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 38]
```diff
@@ -0,0 +1,405 @@
...
+    elif action == "delete":
+        status, message, content, url = delete(module, base_url, headers, index_set_id)
+    elif action == "list":
+        status, message, content, url = list(module, base_url, headers, index_set_id)
+    elif action == "query_index_sets":
+        index_set_id = query_index_sets(module, base_url, headers, title)
+        status, message, content, url = list(module, base_url, headers, index_set_id)
+
+    uresp = {}
+    content = to_text(content, encoding='UTF-8')
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py:487]: This is not required since we are already doing this in individual APIs.

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/monitoring/graylog_index_sets.py (hunk 38 above)
Comment (shortstack) [reply lib/ansible/modules/monitoring/graylog_index_sets.py:487]: are you referring to the `content = to_text(content, encoding='UTF-8')` ?

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 39]
```diff
@@ -0,0 +1,504 @@
...
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            endpoint=dict(type='str'),
+            graylog_user=dict(type='str'),
+            graylog_password=dict(type='str', no_log=True),
+            action=dict(type='str', required=False, default='list', choices=['create', 'update', 'delete', 'list', 'query_index_sets']),
+            title=dict(type='str'),
+            description=dict(type='str'),
+            creation_date=dict(type='str', required=False, default=datetime.datetime.utcnow().isoformat() + 'Z'),
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
            creation_date=dict(type='str', required=False),
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 40]
```diff
@@ -0,0 +1,504 @@
...
+        )
+    )
+
+    endpoint = module.params['endpoint']
+    graylog_user = module.params['graylog_user']
+    graylog_password = module.params['graylog_password']
+    action = module.params['action']
+    title = module.params['title']
+    id = module.params['id']
+    creation_date = datetime.datetime.utcnow().isoformat() + 'Z'
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion
    creation_date = module.params['creation_date'] or datetime.datetime.utcnow().isoformat() + 'Z'
```

File: lib/ansible/modules/monitoring/graylog_index_sets.py [hunk 41]
```diff
@@ -0,0 +1,504 @@
...
+    description:
+      - Description.
+    required: false
+    type: str
+  creation_date:
+    description:
+      - Index set creation date.
+    required: false
+    type: str
+    default: datetime.datetime.utcnow().isoformat() + 'Z'
```
Comment (Akasurde) [inline lib/ansible/modules/monitoring/graylog_index_sets.py]:
```suggestion

```

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Pull request closed without merging**